2-WebApp-graph-user/2-5-HybridFlow/README.md

```diff
@@ -0,0 +1,771 @@
+
+                            return msalInstance.acquireTokenRedirect({
+                                account,
+                                scopes
+                            });
+                        }
+                    });
+            })
+    }
+    else {
+        document.dispatchEvent(new Event('AUTHENTICATED'));
+    }
+})();
+```
+
+## Client-side Graph API requests
+
+In order to make a successful call to Microsoft Graph you first need to get an access token from the `PublicClientApplication` cache and then use that access token within the `Authorization` header along with the `Bearer` key word.
+
+The function `getTokenFromCache` defined within `_Layout.cshtml` does just that.
+
+```JavaScript
+function getTokenFromCache(scopes) {
+    @if (User.Identity is not null)
+    {
+        @Html.Raw($"const username = '{User.Identity.Name}';");
+    }
+    else
+    {
+        @Html.Raw($"const username = '';");
+    }
+
+    const account = msalInstance.getAllAccounts()
+        .find(account => account.username === username);
+
+    return msalInstance.acquireTokenSilent({
+        account,
+        scopes
+    })
+        .catch(error => {
+            if (error instanceof msal.InteractionRequiredAuthError) {
+                return msalInstance.acquireTokenRedirect({
+                    account,
+                    scopes
+                });
+            }
+        });
+};
+```
+
+Actual requests made to the Graph API are handled by the `callMSGraph` function in the `_Layout.cshtml` file.
+
+```JavaScript
+function callMSGraph(path, token) {
+    @{
+        var graphEndpoint = Configuration["DownstreamApi:BaseUrl"];
+        if (!string.IsNullOrEmpty(graphEndpoint))
+        {
+            @Html.Raw($"const graphEndpoint = '{graphEndpoint}';");
+        }
+        else
+        {
+            @Html.Raw($"const graphEndpoint = '';");
+        }
+    }
+
+    const headers = new Headers();
+    const bearer = `Bearer ${token}`;
+    headers.append("Authorization", bearer);
+
+    const options = {
+        method: "GET",
+        headers
+    };
+
+    console.log('request made to Graph API at: ' + new Date().toString());
+
+    return fetch(`${graphEndpoint}${path}`, options)
+        .catch(error => console.log(error))
+}
+```
+
+All requests made by this application to the Graph API are made within the `Profile.cshtml` page.
+
+Before making any requests the JavaScript of the page has a listener for the `AUTHENTICATED` event mentioned earlier. After this event is received requests for the signed in user's profile and contacts is made and rendered into the HTML of the page.
+
+```JavaScript
+async function updateUI() {
+    await Promise.all([
+        loadUserData(),
+        loadUserContacts()
+    ]);
+}
+
+// Only need to run this once to ensure the token has been loaded.
+document.addEventListener('AUTHENTICATED', async () => await updateUI(), { once: true });
+```
```

Comment:
you can skip these sections if you wish. I mean the devs can follow these themselves